### PR TITLE
Example to build custom application and link to libcudf

### LINKED ISSUE #2547: [FEA][DOC] Need examples folder for cudf cpp
Often times when I tell my friends about rapidsai they find that they can't get started on libcudf (C++) without an initial learning curve. I have been told that if we had an examples folder with clear and concise examples of usages of our API that would be a great help. The tests folder is doing a semi decent job at it but because we use internal APIs that are only available in test context they aren't a 100% transferable to an application outside of the project.

### LINKED ISSUE #4824: [DOC] Create example C++ application using cudf APIs
Some users want to integrate existing C++ codebases with cudf but are finding it difficult to get started.

We should create & document an example application that uses libcudf++ to:
1. Read a CSV file
2. Run a filter then groupby and aggregation
3. Write results to CSV

### PR DESCRIPTION
**Pull request by isVoid**
closes #4824, also closes #2547

This PR introduces a basic libcudf example to demonstrate building custom application and link to libcudf using CMake. Libcudf examples now lives under `cpp/examples` folder. There's a `examples/build.sh` that manages building and linking examples. Besides, this PR adds examples into gpuCI tests to make sure examples still builds as the main package evolves.

### CODE REVIEW

**Comment by harrism** [comment]
I don't think we should skip CI. I think we should have simple examples like this inside of the libcudf repo and we should CI them so that we catch breakages.

File: cpp/example/README.md
Comment (harrism) [inline cpp/example/README.md:5]:
```suggestion
This simple C++ example demonstrates a basic libcudf use case and provides a
minimal example of building your own application based on libcudf using CMake.
```

File: cpp/example/README.md
Comment (harrism) [inline cpp/example/README.md:13]: I don't think our minimal example should *require* docker. Users who want to build using cmake on bare metal should be able to build and run the example. Is docker required for this example? If so, can we break it out so it's optional?

File: cpp/example/README.md
Comment (harrism) [inline cpp/example/README.md:1]:
```suggestion
# Basic Standalone libcudf C++ application
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:24]:
```suggestion
  return data_with_meta.tbl;
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:41]: could use a std::transform here.

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:43]:
```suggestion
  return std::make_unique<cudf::table>(std::move(cols));
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:57]: For discussion: do we think that using the memory pool should be part of a minimal libcudf example?  I think just going with the default would be more minimal. We could have a separate example demonstrating the pool.

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:62]:
```suggestion
  // Schema: | Company | Date | Open | High | Low | Close | Volume |
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:67]: I think a view is implicitly constructable from a table, isn't it? In that case you don't need to call `->view()`

```suggestion
  auto keys = get_columns_from_table(table, key_cols);
  auto vals = get_columns_from_table(table, val_cols);
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:70]:
```suggestion
  cudf::groupby::groupby grpby_obj(keys);
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:72]:
```suggestion
  agg_req.values = vals.column(3);
```
?

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:49]:
```suggestion
  size_t available_mem{};
  size_t total_mem{};
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:51]:
```suggestion
  size_t prealloc_mem = available_mem / 2;
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:52]: A comment explaining this line would be nice.

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:84]:
```suggestion
  std::vector<cudf::column_view> cols(res_key->view().begin(), res_key->view().end());
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:77]: Please change all abbreviations "res" to "result". and "reqs" to "requests". Unnecessary abbreviation reduces clarity.

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:70]: Just like you did with read_csv and write_csv, I think it would help clarity to put the core computation in a function, e.g. `average_closing_price()`.

**Review by harrism — CHANGES_REQUESTED** [review]
Thanks for creating this. Mostly just minor style / simplification suggestions.

**Comment by harrism** [comment]
One other thing: I think you should prepare for future libcudf examples, this is not likely to be the only one. So instead of putting it in `/cpp/example`, call the top level directory `examples`, and make a subdirectory for this example, e.g. `/cpp/examples/basic`.

File: cpp/example/Dockerfile
Comment (isVoid) [reply cpp/example/Dockerfile:50]: I'm not very familiar with ccache, but does including it help speed up building libcudf the first time? cc @kkraus14 @trxcllnt (pinging people who may know the answer to my best knowledge)

Comment (harrism) [reply cpp/example/Dockerfile:50]: No, it only helps on subsequent builds.

**Review by ajschmidt8 — APPROVED** [review]

File: cpp/example/README.md
Comment (isVoid) [reply cpp/example/README.md:12]: My guess is yes. Dockerfile is a pretty good description of the dependency needed to build libcudf. I can report back by trying it out on a bare metal machine (virtual machine).

**Comment by harrism** [comment]
@ajschmidt8 already approved for ops, but IMO we should get CI working on this example before it is merged...

File: cpp/example/src/process_csv.cpp
Comment (isVoid) [reply cpp/example/src/process_csv.cpp:57]: Ah this makes sense. For context this example was derived from a performance profiling discussion - there's some left over performance tuning codes.

File: cpp/example/src/process_csv.cpp
Comment (isVoid) [reply cpp/example/src/process_csv.cpp]: I realized in this example (getting two columns from a table), `get_column_from_table` seems excessive. Removed.

File: cpp/example/CMakeLists.txt
Comment (isVoid) [reply cpp/example/CMakeLists.txt:26]: Currently `main` branch is mapped to stable release, which is `branch-0.18` I believe? I think the CPM goodness was not introduced after https://github.com/rapidsai/cudf/pull/7107. Once `branch-0.19` becomes main we should be able to just use `main` tag.

**Comment by isVoid** [comment]
A few updates:
- Code is more concise, removed helper function `get_columns_from_table`
- Removed memory pool init
- Removed ccache

**Comment by isVoid** [comment]
Adding [skip-ci] to avoid wasting CI resource before we decide how to test this example.

File: cpp/example/Dockerfile
Comment (harrism) [inline cpp/example/Dockerfile:46]: Not needed anymore?

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:20]:
```suggestion
  return data_with_meta.tbl;
```

File: cpp/example/src/process_csv.cpp
Comment (harrism) [inline cpp/example/src/process_csv.cpp:42]:
```suggestion
  cudf::groupby::aggregation_request agg_request{val, {cudf::make_mean_aggregation()}};
  std::vector<cudf::groupby::aggregation_request> requests{std::move(agg_request)};
```

**Review by harrism — CHANGES_REQUESTED** [review]
The example is much simpler, good job. Couple more simplifications.

File: cpp/example/src/process_csv.cpp
Comment (ttnghia) [inline cpp/example/src/process_csv.cpp:13]: I think we should follow `cudf` convention for header inclusion: include from near to far (https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/docs/DEVELOPER_GUIDE.md#includes). So, include the headers from `cudf` first, then `rmm`, and finally from std.

**Review by ttnghia — CHANGES_REQUESTED** [review]

File: cpp/example/src/process_csv.cpp
Comment (isVoid) [reply cpp/example/src/process_csv.cpp:20]: I think `move` is required here because the struct `data_with_meta` still owns the the `unique_ptr`. Similar to why we need a move for a `unique_ptr` in a vector.
example: https://godbolt.org/z/T9x5qWfYa

File: cpp/example/src/process_csv.cpp
Comment (isVoid) [reply cpp/example/src/process_csv.cpp:42]: I tried this, but getting compile error:
```
In file included from /usr/include/c++/9/memory:65,
                 from /workspace/build/_deps/cudf-src/cpp/include/cudf/aggregation.hpp:22,
                 from /workspace/src/process_csv.cpp:1:
/usr/include/c++/9/bits/stl_uninitialized.h: In instantiation of '_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = const std::unique_ptr<cudf::aggregation>*; _ForwardIterator = std::unique_ptr<cudf::aggregation>*]':
/usr/include/c++/9/bits/stl_uninitialized.h:307:37:   required from '_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = const std::unique_ptr<cudf::aggregation>*; _ForwardIterator = std::unique_ptr<cudf::aggregation>*; _Tp = std::unique_ptr<cudf::aggregation>]'
/usr/include/c++/9/bits/stl_vector.h:1582:33:   required from 'void std::vector<_Tp, _Alloc>::_M_range_initialize(_ForwardIterator, _ForwardIterator, std::forward_iterator_tag) [with _ForwardIterator = const std::unique_ptr<cudf::aggregation>*; _Tp = std::unique_ptr<cudf::aggregation>; _Alloc = std::allocator<std::unique_ptr<cudf::aggregation> >]'
/usr/include/c++/9/bits/stl_vector.h:626:2:   required from 'std::vector<_Tp, _Alloc>::vector(std::initializer_list<_Tp>, const allocator_type&) [with _Tp = std::unique_ptr<cudf::aggregation>; _Alloc = std::allocator<std::unique_ptr<cudf::aggregation> >; std::vector<_Tp, _Alloc>::allocator_type = std::allocator<std::unique_ptr<cudf::aggregation> >]'
/workspace/src/process_csv.cpp:41:86:   required from here
/usr/include/c++/9/bits/stl_uninitialized.h:127:72: error: static assertion failed: result type must be constructible from value type of input range
  127 |       static_assert(is_constructible<_ValueType2, decltype(*__first)>::value,
      |                                                                        ^~~~~
CMakeFiles/libcudf_example.dir/build.make:84: recipe for target 'CMakeFiles/libcudf_example.dir/src/process_csv.cpp.o' failed
```

File: cpp/example/src/process_csv.cpp
Comment (jrhemstad) [inline cpp/example/src/process_csv.cpp:57]: This isn't necessary. `get_current_device_resource` returns a `cuda_memory_resource*` by default if one was never set.

File: cpp/example/src/process_csv.cpp
Comment (isVoid) [reply cpp/example/src/process_csv.cpp:42]: Min-repro: https://godbolt.org/z/d6Pqb34ed

**Comment by isVoid** [comment]
Updates:
- Further reduced some dependencies in dockerfile
- Tested on building the project on local machine without docker environment, readme updated accordingly.
- Files organized into examples/basic

File: cpp/examples/basic/README.md
Comment (isVoid) [inline cpp/examples/basic/README.md:16]: I'm not sure if it's this project's responsibility to account for prerequisites here or maybe I should let users refer to `Contributing.md`?

File: cpp/examples/basic/Dockerfile
Comment (trxcllnt) [inline cpp/examples/basic/Dockerfile:2]: Should move this down to line 35 to avoid re-running all the `apt install` if someone changes this build arg.

File: cpp/example/src/process_csv.cpp
Comment (harrism) [reply cpp/example/src/process_csv.cpp]: https://github.com/rapidsai/cudf/blob/500f42c16f19ed475569b9bf2d7055a3abfc25ed/cpp/tests/groupby/groupby_test_util.hpp#L73-L74

Perhaps like this then:

```
std::vector<cudf::groupby::aggregation_request> requests;
requests.emplace_back(val, {cudf::make_mean_aggregation());
```

File: cpp/examples/basic/README.md
Comment (harrism) [reply cpp/examples/basic/README.md:16]: Well, the prereqs are different if they are relying on a pre-built libcudf.

File: cpp/example/src/process_csv.cpp
Comment (isVoid) [reply cpp/example/src/process_csv.cpp]: I think there are extra two lines below that are also relevant in creating an aggregation. For it's complexity, I created a separate function `make_single_aggregation_request` to wrap the logic.

An offline discussion mentions all these complexity maybe alleviated once the vectors are replaced with `host_span`. And that APIs like `aggregation_request` shouldn't be owner to an `aggregation_request`.

File: cpp/examples/basic/Dockerfile
Comment (isVoid) [reply cpp/examples/basic/Dockerfile:2]: Could you clarify, is it just line 2 or both line 1, 2? Can't quite tell from gh's UI.

**Comment by isVoid** [comment]
A second thought: this example should only demonstrate "basic libcudf use case and how to build a custom application on top of libcudf". The former is shown by `process_csv.cpp` and the latter is shown with `CMakeLists.txt`. Providing a `Dockerfile` to establish an environment for building seems unrelated, and adds difficulty for maintenance, thus removed.

Added readme at root level of `examples` to show current list of examples.

For build scripts, AFAIK there can be two ways:
- Add examples build scripts in cudf root `build.sh`
- A separate build.sh in `examples`, call it from root `build.sh`
Which better?

File: cpp/example/CMakeLists.txt
Comment (isVoid) [reply cpp/example/CMakeLists.txt]: Now linked to `main` branch.

**Comment by isVoid** [comment]
rerun tests

File: cpp/examples/basic/README.md
Comment (harrism) [inline cpp/examples/basic/README.md:20]: Could mention they can alternatively build using `build.sh`

File: cpp/examples/basic/src/process_csv.cpp
Comment (harrism) [inline cpp/examples/basic/src/process_csv.cpp:17]: I don't think the move is necessary here.  Can just collapse these two lines: 
```suggestion
  return cudf::io::read_csv(options);
```

**Review by harrism — CHANGES_REQUESTED** [review]
This is looking great. One tiny change.

File: cpp/examples/basic/src/process_csv.cpp
Comment (isVoid) [reply cpp/examples/basic/src/process_csv.cpp]: So `read_csv` returns a `table_with_metadata` struct, which is the owner of the table. We could instead return the `data_with_meta` (which I think that's pretty neat as well). Updated with that.

**Comment by isVoid** [comment]
Example buildings are added to gpuCI tests. For completeness, we should probably execute the built binaries too. However each example may have their own file dependencies (e.g. the basic example requires I/O to a csv file).

**Comment by harrism** [comment]
> Example buildings are added to gpuCI tests. For completeness, we should probably execute the built binaries too. However each example may have their own file dependencies (e.g. the basic example requires I/O to a csv file). Any idea how to best organize the executables?

We don't do this for benchmarks, so I think it's OK to just build them for now.

File: cpp/examples/basic/src/process_csv.cpp
Comment (harrism) [reply cpp/examples/basic/src/process_csv.cpp]: Sorry, I read the `.tbl` as `_tbl` and missed that you were extracting a field of a struct.

Comment (harrism) [reply cpp/examples/basic/src/process_csv.cpp]: I think part of the problem is you use `auto` everywhere so the types are not obvious. That's fine, but in examples sometimes it's nice to see the types to aid learning the API.

**Review by harrism — APPROVED** [review]
Nice example. Thanks.

File: cpp/examples/build.sh
Comment (ttnghia) [inline cpp/examples/build.sh:10]: I'm not an expert in cmake thus I'm not sure if this will result in `-j 4` or `-j -4`?

**Review by ttnghia — APPROVED** [review]

File: cpp/examples/build.sh
Comment (harrism) [reply cpp/examples/build.sh:10]: Seems to be the same as here: https://github.com/rapidsai/cudf/blob/0b9f1789e5d5bec6a38fe57c7e4556e0dc359bdf/ci/gpu/build.sh#L17

**Comment by harrism** [comment]
@gpucibot merge

**Comment by harrism** [comment]
rerun tests

File: cpp/examples/build.sh
Comment (ttnghia) [reply cpp/examples/build.sh:10]: I got it! We will have `4`, not `-4`. The shell syntax is a bit confusing:
```
${parameter:-word}
    If parameter is unset or null, the expansion of word is substituted. 
    Otherwise, the value of parameter is substituted.
```

**Comment by isVoid** [comment]
@harrism should this rebase to `21.08`?

**Comment by harrism** [comment]
Oh, OK. I thought it was fine to go in 21.06.

**Comment by galipremsagar** [comment]
rerun tests

**Comment by davidwendt** [comment]
Would this also close issue #2547 ?

**Comment by isVoid** [comment]
rerun tests

**Comment by isVoid** [comment]
Looks like the failed CI builds are due to mis-configuration of paths in `build.sh`. Investigating.

File: cpp/examples/basic/CMakeLists.txt
Comment (isVoid) [inline cpp/examples/basic/CMakeLists.txt:5]: This should be 17 now I guess?
```suggestion
set(CMAKE_CXX_STANDARD 17)
```

File: cpp/examples/basic/CMakeLists.txt
Comment (isVoid) [inline cpp/examples/basic/CMakeLists.txt:21]: Hmm, so [this build issue](https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-v0.20/job/cudf/job/prb/job/cudf-gpu-test/CUDA=11.0,GPU_LABEL=gpu,OS=ubuntu16.04,PYTHON=3.7/1275/console) probably result from a mismatched versioning of dlpack, which was updated in `21.06` at this PR: #8262 . But this example app still attempts to build the `main` (19.02) branch in the CI environment.

The purpose of adding the examples to CI test is to test if any subsequent PRs will break the examples. But currently this only tests whether the CI test builds with the latest release. I can foresee this will bring in a lot of problem at the end of every release (when `main` branch is updated).

I'm exploring a flag/argument for `CMake` that can dynamically build/link to the `current actively developing branch` in CI environment, while keeping it defaulted to `main` outside of CI so that libcudf explorers can still access the stable release. Other ideas are also welcome.

cc @harrism

**Comment by isVoid** [comment]
Thanks to @robertmaynard pointing out, that the examples should not build from `main`, because older examples may break with latest future releases. Instead, they should link to the latest version. This requires auto bumping the version from `update-version.sh`, which is updated accordingly.

Requires re-review from @rapidsai/ops-codeowners

**Comment by isVoid** [comment]
Oops. I forgot to block merging. @robertmaynard can you still look at the CMake file and raise an issue if there's anything wrong with it?

**Comment by robertmaynard** [comment]
In my mind examples should have as little 'noise' so that readers can easily find
the important bits. To this end we would want to do the following:

Refactor the `add_executable` and `target_link_libaries` lines to look like:
```
add_executable(basic_example src/process_csv.cpp)
target_link_libraries(basic_example PRIVATE cudf::cudf)
```

This has multiple improvements:
 - We use real names instead of variables, making it easier to read
 - We don't quote relative paths, since that is uneeded
 - We use the Modern CMake pattern of marking `target_link_libraries` libraries
   as public/private/interface.

Drop the `set(CMAKE_CUDA_ARCHITECTURES "")` line. The can be specified in the
`build.sh`. If we are wanting a nicer API for specifying which cuda archs we are
building for we could use `rapids_cmake`

Drop the `set(CMAKE_EXPORT_COMPILE_COMMANDS ON)` entirely. This is only needed for
some autocomplete tools.

Drop the `set(CMAKE_CXX_STANDARD 17)` line and instead at the end of the file do:
```
target_compile_features(basic_example PRIVATE cxx_std_17)
```


The `project(basic_example` is enabling the CUDA and C languages which aren't
needed to use `cudf` so we should drop them. If `cudf` is subsequently built
from source via CPM it will enable the `CUDA` language when needed


I think we can clean up the logic around downloading and using CPM, it currently has some logic that really only makes sense for 'production' projects. Plus it is using an old CPM version fetched from the old project URL.

I think we can simplify the cpm logic to be:

```
set(CPM_DOWNLOAD_VERSION v0.32.2)
file(DOWNLOAD https://github.com/cpm-cmake/CPM.cmake/releases/download/${CPM_DOWNLOAD_VERSION}/get_cpm.cmake ${CMAKE_BINARY_DIR}/cmake/get_cpm.cmake)
include(${CMAKE_BINARY_DIR}/cmake/get_cpm.cmake)

set(CUDF_VERSION "21.08")
CPMFindPackage(NAME  cudf
    GIT_REPOSITORY  https://github.com/rapidsai/cudf
    GIT_TAG         branch-${VERSION}
    GIT_SHALLOW     TRUE
    SOURCE_SUBDIR   cpp
)
```

Total end result would be the CMakeLists.txt looking like:

```
cmake_minimum_required(VERSION 3.18)

project(basic_example VERSION 0.0.1 LANGUAGES C CXX CUDA)

set(CPM_DOWNLOAD_VERSION v0.32.2)
file(DOWNLOAD https://github.com/cpm-cmake/CPM.cmake/releases/download/${CPM_DOWNLOAD_VERSION}/get_cpm.cmake ${CMAKE_BINARY_DIR}/cmake/get_cpm.cmake)
include(${CMAKE_BINARY_DIR}/cmake/get_cpm.cmake)

set(CUDF_TAG "branch-21.08")
CPMFindPackage(NAME  cudf
    GIT_REPOSITORY  https://github.com/rapidsai/cudf
    GIT_TAG         ${CUDF_TAG}
    GIT_SHALLOW     TRUE
    SOURCE_SUBDIR   cpp
)

# Configure your project here
add_executable(basic_example src/process_csv.cpp)
target_link_libraries(basic_example PRIVATE cudf::cudf)
target_compile_features(basic_example PRIVATE cxx_std_17)

```

**Pull request merged**
